### PR TITLE
Optional live video

### PR DESCRIPTION
**Pull request by korabelnikov**
facenets/backend/print_face_info_on_image.py  это переименованный и измененный  facenets/backend/facenet_image.py

### CODE REVIEW

File: facenets/backend/image_to_jpeg_bytes.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+from io import BytesIO
```
Comment (rozhn0v) [inline facenets/backend/image_to_jpeg_bytes.py]: i think this module should be renamed to something like "image_converter"

File: facenets/backend/observer_server.py [hunk 2]
```diff
@@ -8,36 +8,46 @@
 
 log = logging.getLogger(__name__)
 
-_CLIENTS = {}
+web_clients = {}
```
Comment (rozhn0v) [inline facenets/backend/observer_server.py]: why?

File: facenets/backend/print_face_info_on_image.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
+import cv2
```
Comment (rozhn0v) [inline facenets/backend/print_face_info_on_image.py]: i think this module should be renamed too - not verb-based

File: facenets/backend/print_face_info_on_image.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+                  thickness=1)
+    cv2.putText(im,
+                '{}. distance {:.2f}'.format(name, distance),
+                (y1, x1),
+                *_USUAL_ARGS)
+
+
+def print_face_info_on_image(im, faces, names_and_distances):
+    assert len(faces) == len(names_and_distances), "this collections have to have corresponding elements"
+    if len(faces) == 0:
```
Comment (rozhn0v) [inline facenets/backend/print_face_info_on_image.py]:
```
if faces:
    # foo
```
should be enough

File: facenets/backend/print_face_info_on_image.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+                  (255, 0, 0),
+                  thickness=1)
+    cv2.putText(im,
+                '{}. distance {:.2f}'.format(name, distance),
+                (y1, x1),
+                *_USUAL_ARGS)
+
+
+def print_face_info_on_image(im, faces, names_and_distances):
+    assert len(faces) == len(names_and_distances), "this collections have to have corresponding elements"
```
Comment (rozhn0v) [inline facenets/backend/print_face_info_on_image.py]: please, check flake/pylint warnings for this module

File: facenets/backend/run_server.py [hunk 6]
```diff
@@ -41,12 +41,16 @@ def _setup_logger():
...
 @app.route('/')
 def index():
-    # clear all listeners TODO it's weird for multisession
+    # clear all listeners TODO? it's weird for multisession
     clients = CAMERA_SERVER.clients()
     CAMERA_SERVER.remove_all_listeners()
     BROADCASTING_KEYS.clear()
     return render_template('index.html', cams=clients)
 
+#@app.route('/realtime_face_log')
```
Comment (rozhn0v) [inline facenets/backend/run_server.py]: this should be removed

File: facenets/backend/restream.py [hunk 7]
```diff
@@ -1,21 +1,53 @@
...
+        response = {'results': results}
+
+        # send response for such web clients who don't want live video
+        ws.send_data_to_observers_of_camera(self._camera_address, json.dumps(response),
+                                            lambda c: not c.is_live_streaming())
 
-        for coordinates in face_coordinates:
-            image.tag_face(coordinates)
+        # and send if any of web clients want live video
+        is_live_streaming = any((c.is_live_streaming() for c in ws.web_clients))
```
Comment (rozhn0v) [inline facenets/backend/restream.py]: move this logic into observer_server module, make a function

File: facenets/backend/observer_server.py [hunk 8]
```diff
@@ -8,36 +8,46 @@
...
 
 
-def send_data_to_observers_of_camera(target_camera_address, data):
-    for web_socket, camera_address in _CLIENTS.items():
+def send_data_to_observers_of_camera(target_camera_address, data, web_client_condition_fn=None):
```
Comment (rozhn0v) [inline facenets/backend/observer_server.py]: i think it's better to have two separate methods (for text data clients and for video data clients) than to pass a flag

File: facenets/backend/image_to_jpeg_bytes.py (hunk 1 above)
Comment (korabelnikov) [reply facenets/backend/image_to_jpeg_bytes.py]: done

File: facenets/backend/observer_server.py (hunk 2 above)
Comment (korabelnikov) [reply facenets/backend/observer_server.py]: I used it in different module to determine which client wants live video and which doesn't want

File: facenets/backend/print_face_info_on_image.py (hunk 3 above)
Comment (korabelnikov) [reply facenets/backend/print_face_info_on_image.py]: I've hanged to different one, that's better. But it's still verb-based

File: facenets/backend/print_face_info_on_image.py (hunk 4 above)
Comment (korabelnikov) [reply facenets/backend/print_face_info_on_image.py]: It wasn't enough if faces is np.array([x1,y1,x2,y2]).
See error "True value of ndarray is ambiguous (use any() of all()"

File: facenets/backend/print_face_info_on_image.py (hunk 5 above)
Comment (korabelnikov) [reply facenets/backend/print_face_info_on_image.py]: done

File: facenets/backend/run_server.py (hunk 6 above)
Comment (korabelnikov) [reply facenets/backend/run_server.py]: done

File: facenets/backend/restream.py (hunk 7 above)
Comment (korabelnikov) [reply facenets/backend/restream.py]: I think this is not a responsibility of observer_server

File: facenets/backend/observer_server.py (hunk 8 above)
Comment (korabelnikov) [reply facenets/backend/observer_server.py]: done

File: tests/restreamer_test.py [hunk 9]
```diff
@@ -1,37 +1,28 @@
...
 from unittest.mock import Mock
-import itertools
+import facenets.backend.put_faces_on_image as processor
 
-from facenets.backend.restream import Restreamer
 
+# TODO get rid of text literals. Maybe use context manager
+# moreover @unittest.mock.patch should be in inverse order
 
 class RestreamerTest(unittest.TestCase):
```
Comment (rozhn0v) [inline tests/restreamer_test.py]: rename this and file according to tested module

File: tests/restreamer_test.py [hunk 10]
```diff
@@ -1,37 +1,28 @@
...
-        detector.get_faces.return_value = itertools.chain([Mock()])
-
-        image = Mock()
 
-        restreamer = Restreamer(
-            camera_address=Mock(),
-            detector=detector,
-            image_factory=Mock(return_value=image))
-        restreamer.restream(image_array=Mock())
+    @unittest.mock.patch('facenets.backend.print_face_info_on_image._tag_face')
```
Comment (rozhn0v) [inline tests/restreamer_test.py]: rename variables and use parameters with args where Mocks are passed

File: facenets/ml/detector.py [hunk 11]
```diff
@@ -2,34 +2,33 @@
 import os
```
Comment (rozhn0v) [inline facenets/ml/detector.py]: just revert to master version

File: facenets/backend/restream.py [hunk 12]
```diff
@@ -1,21 +1,54 @@
...
+
+        # send response for such web clients who don't want live video
+        ws.send_to_text_observers(self._camera_address,
+                                  json.dumps(response))
 
-        for coordinates in face_coordinates:
-            image.tag_face(coordinates)
+        # and send if any of web clients want live video
+        is_live_streaming = any(
+            (c.is_live_streaming() for c in ws.web_clients))
```
Comment (rozhn0v) [inline facenets/backend/restream.py]: move to observer_clients as a function

File: facenets/backend/put_faces_on_image.py [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+                *_USUAL_ARGS)
+
+
+def print_face_info_on_image(im, faces, names_and_distances):
+    faces = list(faces)
+    names_and_distances = list(names_and_distances)
+    assert len(
+        faces) == len(names_and_distances
+                      ), "this collections have to have corresponding elements"
+    if len(faces) == 0:
```
Comment (rozhn0v) [inline facenets/backend/put_faces_on_image.py]: leave a comment about this check (and TODO for unit test)

File: facenets/backend/observer_server.py [hunk 14]
```diff
@@ -8,36 +8,61 @@
...
+                                         lambda c: c.is_live_streaming())
+
+
+def send_to_text_observers(target_camera_address, data):
+    send_data_to_all_observers_of_camera(target_camera_address,
+                                         data,
+                                         lambda c: not c.is_live_streaming())
+
+
+def send_data_to_all_observers_of_camera(target_camera_address,
```
Comment (rozhn0v) [inline facenets/backend/observer_server.py]: remove lambdas, make this function package-protected

File: tests/restreamer_test.py (hunk 9 above)
Comment (korabelnikov) [reply tests/restreamer_test.py]: done

File: tests/restreamer_test.py (hunk 10 above)
Comment (korabelnikov) [reply tests/restreamer_test.py]: done

File: facenets/ml/detector.py (hunk 11 above)
Comment (korabelnikov) [reply facenets/ml/detector.py]: done

File: facenets/backend/restream.py (hunk 12 above)
Comment (korabelnikov) [reply facenets/backend/restream.py]: done

File: facenets/backend/put_faces_on_image.py (hunk 13 above)
Comment (korabelnikov) [reply facenets/backend/put_faces_on_image.py]: done

**Pull request merged**